examples/chat/src/routes/messages.rs

```diff
@@ -0,0 +1,164 @@
+    }
+
+    // Acquire a database connection and execute the insert.
+    let message = {
+        let pool = head.state().pool();
+        let mut conn = pool.get().await?;
+
+        insert.get_result(&mut conn).await?
+    };
+
+    Response::build().status(201).json(&message)
+}
+
+pub async fn show(request: Request, _: Next) -> via::Result {
+    // Preconditions
+    let id = request.param("message-id").parse()?;
+
+    // Build the query from URI params.
+    let query = Message::select().filter(by_id(&id));
+
+    // Print the query to stdout in debug mode.
+    if cfg!(debug_assertions) {
+        println!("\n{}", diesel::debug_query::<Pg, _>(&query));
+    }
+
+    // Acquire a database connection and execute the query.
+    let message: MessageWithJoins = {
+        let pool = request.state().pool();
+        let mut conn = pool.get().await?;
+
+        query.first(&mut conn).await?
+    };
+
+    Response::build().json(&message)
+}
+
+pub async fn update(request: Request, _: Next) -> via::Result {
+    // Preconditions
+    let current_user_id = request.current_user()?.id;
+    let message_id = request.param("message-id").parse()?;
+
+    // Deserialize the request body into message params.
+    let (head, future) = request.into_future();
+    let change_set = future.await?.serde_json::<MessageChangeSet>()?;
+
+    // Build the update statement with the params from the body.
+    let update = diesel::update(Message::TABLE)
+        .set(change_set)
+        // Proceed if the message is authored by the current user.
+        .filter(by_id(&message_id).and(by_author(current_user_id)))
+        .returning(Message::as_returning());
+
+    // Print the query to stdout in debug mode.
+    if cfg!(debug_assertions) {
+        println!("\n{}", diesel::debug_query::<Pg, _>(&update));
+    }
+
+    // Acquire a database connection and execute the update.
+    let message = {
+        let pool = head.state().pool();
+        let mut conn = pool.get().await?;
+
+        update.get_result(&mut conn).await.found_or_forbidden()?
+    };
+
+    Response::build().json(&message)
+}
+
+pub async fn destroy(request: Request, _: Next) -> via::Result {
+    // Preconditions
+    let current_user_id = request.current_user()?.id;
+    let message_id = request.param("message-id").parse()?;
+
+    // Build the delete statement from URI params.
+    let delete = diesel::delete(Message::TABLE).filter(
+        // Proceed if the message is authored by the current user.
+        by_id(&message_id).and(by_author(current_user_id)),
+    );
+
+    // Print the query to stdout in debug mode.
+    if cfg!(debug_assertions) {
+        println!("\n{}", diesel::debug_query::<Pg, _>(&delete));
+    }
+
+    // Acquire a database connection and execute the delete.
+    {
+        let pool = request.state().pool();
+        let mut conn = pool.get().await?;
+
+        delete.execute(&mut conn).await.found_or_forbidden()?;
```

Comment:
I really like this pattern for access control. It doesn't fail early and allows for human error but it works exactly as you would expect without any extra database queries. A good reason for logging the query in debug mode.